### ISSUE TITLE
Conflict between t:jsDecode and t:cssDecode

### ISSUE THREAD

**Issue by Olament** [SpiderLabs/ModSecurity#2517]
**Describe the bug**

Using ```t:jsDecode``` and ```t:cssDecode``` together in a transformation sequence is a common practice in CRS (ex. APPLICATION-ATTACK-XXS). However, placing ```t:jsDecode``` before ```t:cssDecode``` may cause ModSecurity's failure to decode CSS-encoded string under some circumstance. 

This is because both JS-encoded string and CSS-encoded string use "\\" as the escaping character. When a CSS-encoded string with form ```\HH```, where H is valid hexadecimal character, went through the transformation sequence, it will be interpreted as ```\C``` form by ```t:jsDecode``` (see [js_decode.cc](https://github.com/SpiderLabs/ModSecurity/blob/50fc347ed4d0852d54561e15559de07c698fbb16/src/actions/transformations/js_decode.cc#L116)).  Since ```\HH``` is not any of these escaped whitespace character, ```t:jsDecode``` will transform ```\HH``` into ```HH```, which cause problem for later ```t:cssDecode``` transformation .

**To Reproduce**

***When using ```t:cssDecode``` alone***
```
SecRule REQUEST_HEADERS:User-Agent "@rx jkl" \
    "id:941110,\
    phase:2,\
    deny,\
    t:none,t:jsDecode,t:cssDecode"
```
```
# Get status code 403, work as expected
curl --request GET \
  --url http://ModSec \
  --header 'User-Agent: \6A\6B\6C' \
```
```
# Get status code 403, work as expected
curl --request GET \
  --url http://ModSec \
  --header 'Content-Type: multipart/form-data' \
  --header 'User-Agent: jkl' \
```
***When using ```t:jsDecode``` before ```t:cssDecode```***
```
SecRule REQUEST_HEADERS:User-Agent "@rx jkl" \
    "id:941110,\
    phase:2,\
    deny,\
    t:none,t:jsDecode,t:cssDecode"
```
```
# Get HTTP status code 200, Expect 403
curl --request GET \
  --url http://ModSec \
  --header 'User-Agent: \6A\6B\6C' \
```
```
# Get status code 403, work as expected
curl --request GET \
  --url http://ModSec \
  --header 'Content-Type: multipart/form-data' \
  --header 'User-Agent: jkl' \
```
***When using ```t:cssDecode``` before ```t:jsDecode```***
```
SecRule REQUEST_HEADERS:User-Agent "@rx jkl" \
    "id:941110,\
    phase:2,\
    deny,\
    t:none,t:cssDecode,t:jsDecode"
```
```
# Get status code 403, work as expected
curl --request GET \
  --url http://ModSec \
  --header 'User-Agent: \6A\6B\6C' \
```
```
# Get status code 403, work as expected
curl --request GET \
  --url http://ModSec \
  --header 'Content-Type: multipart/form-data' \
  --header 'User-Agent: jkl' \
```

**Additional context**

Other transformation that use "\\" to mark the start of escape sequence may having the same problem.

**Comment by martinhsv** [comment]
Hi @Olament ,

Each transformation is a specific tool with certain uses.  And it is, of course, possible to use them in sequence.

But that doesn't mean that all combinations are sound, or that all sequences are sound.  And it definitely doesn't mean that all permutations are sound in all circumstances.

Ideally, a ModSecurity rule should 'see' the data in the same way that the end application will see it.  This is often an easier task with a custom rule because one can craft the set of transformations to match what the endpoint does.  This is a bigger challenge with a generic rule simply because it is often trying to emulate a variety of potential endpoint data interpretations.

I don't think I see an issue with ModSecurity per se (unless you're asserting that jsDecode's output is a bug -- which it doesn't seem like you are).  Overall it's the responsibility of the rule writer to ensure that they are selecting the proper transformations and placing them in a sensible order.  You may wish to raise this with the crs maintainers as they may have thoughts about this rule pattern of theirs.

If 't:jsDecode,t:cssDecode' is a fundamentally illogical transformation sequence, we could look to add a warning.  (We are generally moving towards providing warnings about logically flawed rules -- e.g. a rule that attempts to inspect POST_ARGS in phase:1).)  However, I haven't thought through that 't:jsDecode,t:cssDecode' is in that category.

If it is not a sequence that is in the 'always illogical' category, did you have another possibility in mind as to what the engine's behaviour should be?

**Comment by Olament** [comment]
Thanks for the detailed reply @martinhsv 

I think one of the solution is to swap the order of ```t:jsDecode``` and ```t:cssDecode``` in CRS. 

Another solution would be changing the implementation of ```t:jsDecode``` such that it will return the consumed backslash "\\" if the following character sequence does not form a valid encoding. For input string "\6A\6B\6C", current implementation will consume all blackslash and return "6A6B6C". If we add the valid encoding check, ```t:jsDecode``` will first check if "\6" is a valid js escaped sequence and, since it is not, return "\6A\6B\6C" instead. However, I am not sure if returning the consumed backslash follows the JavaScript encoding/decoding standard.

**Comment by martinhsv** [comment]
In this case, however, it _is_ a valid encoding.

Octal escape sequences have a backslash followed by one, two or three octal digits.

The two-character sequence "\6" is a valid octal escape sequence encoding of '6'

When the jsDecode transformation is happening, the code sees "\6" which so far is a valid octal escape sequence.  The next character is not an octal digit, which means that the the "\6" octal escape sequence is complete, and the new character ('A') is new content.

The jsDecode thus turns the 9-character "\6A\6B\6C" into a 6-byte string where the 2nd, 4th and 6th bytes are 'A', 'B' and 'C' respectively, while the 1st, 3rd and 5th bytes are each the hexadecimal digit 0x06:  "[0x06]A[0x06]B[0x06]C".

In other words this is not a case where "the following character sequence does not form a valid encoding"

**Comment by Olament** [comment]
You're right! Thanks for the explanation.
